ql/cashflows/overnightindexedcoupon.cpp

```diff
@@ -83,25 +87,38 @@ namespace QuantLib {
                     }
                 }
 
+                auto endIndex = i;
+
+                if (minDate < valueDates[n]) {
+                    while (endIndex < n && minDate <= valueDates[endIndex])
+                        endIndex++;
+                } else {
+                    endIndex = n;
+                }
+
                 // forward part using telescopic property in order
                 // to avoid the evaluation of multiple forward fixings
-                if (i<n) {
-                    Handle<YieldTermStructure> curve =
+                if (i < endIndex) {
+                    const Handle<YieldTermStructure> curve =
                         index->forwardingTermStructure();
                     QL_REQUIRE(!curve.empty(),
                                "null term structure set to this instance of "<<
                                index->name());
 
-                    const vector<Date>& dates = coupon_->valueDates();
-                    DiscountFactor startDiscount = curve->discount(dates[i]);
-                    DiscountFactor endDiscount = curve->discount(dates[n]);
+                    const DiscountFactor startDiscount = curve->discount(valueDates[i]);
+                    const DiscountFactor endDiscount = curve->discount(valueDates[endIndex]);
 
                     compoundFactor *= startDiscount/endDiscount;
                 }
 
-                Rate rate = (compoundFactor - 1.0) / coupon_->accrualPeriod();
+                const Rate rate = (compoundFactor - 1.0) / coupon_->accrualPeriod();
```

Comment:
This should be divided by the period between `accrualStartDate_` and `date`, not the whole accrual period for the coupon.